driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -185,25 +185,27 @@ public long getMaxAwaitTimeMS() {
         return timeoutSettings.getMaxAwaitTimeMS();
     }
 
-    public void runMaxTimeMSTimeout(final Runnable onInfinite, final LongConsumer onRemaining,
-            final Runnable onExpired) {
+    public void runMaxTimeMS(final LongConsumer onRemaining) {
         if (maxTimeSupplier != null) {
-            runWithFixedTimout(maxTimeSupplier.get(), onInfinite, onRemaining);
+            runWithFixedTimout(maxTimeSupplier.get(), onRemaining);
             return;
         }
-
-        if (timeout != null) {
-            timeout.shortenBy(minRoundTripTimeMS, MILLISECONDS)
-                    .run(MILLISECONDS, onInfinite, onRemaining, onExpired);
-        } else {
-            runWithFixedTimout(timeoutSettings.getMaxTimeMS(), onInfinite, onRemaining);
+        if (timeout == null) {
+            runWithFixedTimout(timeoutSettings.getMaxTimeMS(), onRemaining);
+            return;
         }
+        timeout.shortenBy(minRoundTripTimeMS, MILLISECONDS)
+                .run(MILLISECONDS,
+                        () -> {},
+                        onRemaining,
+                        () -> {
+                            throw createMongoRoundTripTimeoutException();
+                        });
+
     }
 
-    private static void runWithFixedTimout(final long ms, final Runnable onInfinite, final LongConsumer onRemaining) {
-        if (ms == 0) {
-            onInfinite.run();
-        } else {
+    private static void runWithFixedTimout(final long ms, final LongConsumer onRemaining) {
```

Comment:
Nit: There was a typo in the original name: 

`runWithFixedTimout` => `runWithFixedTimeout`

Reply:
Done, thanks!